File: tests/initdata.py
#!/usr/bin/env python3
"""
Telegram Init Data Generator
Generates mock Telegram WebApp init data with valid hash signature.
"""

import hmac
import hashlib
import urllib.parse
import json
import time
import os
from typing import Dict, Any

# Mock data configuration
MOCK_USER_DATA = {
    "id": 687627953,
    "first_name": "linuxfight",
    "last_name": "",  # Empty last name
    "username": "linuxfight",
    "language_code": "ru",
    "is_premium": False,
    "allows_write_to_pm": False,
    "photo_url": "https://example.com/cat.jpeg"
}

def load_bot_token() -> str:
    """
    Load bot token from dev.env file.
    Expected format in dev.env: BOT_TOKEN=your_token_here
    """
    env_file = "dev.env"

    if not os.path.exists(env_file):
        raise FileNotFoundError(f"{env_file} file not found. Please create it with BOT_TOKEN=your_token")

    with open(env_file, 'r') as file:
        for line in file:
            line = line.strip()
            if line.startswith('BOT_TOKEN='):
                token = line.split('=', 1)[1].strip()
                if not token:
                    raise ValueError("BOT_TOKEN is empty in dev.env file")
                return token

    raise ValueError("BOT_TOKEN not found in dev.env file")

def create_user_json(user_data: Dict[str, Any]) -> str:
    """
    Create user JSON string for Telegram init data.
    Removes empty string values to match Telegram's behavior.
    """
    # Remove empty string values (but keep False boolean values)
    filtered_data = {
        k: v for k, v in user_data.items()
        if v != "" and v is not None
    }

    # Convert to JSON without spaces (compact format)
    return json.dumps(filtered_data, separators=(',', ':'))

def generate_init_data_hash(data_dict: Dict[str, str], bot_token: str) -> str:
    """
    Generate hash for Telegram init data using the official algorithm:
    1. Create array of key=value pairs (excluding hash)
    2. Sort alphabetically
    3. Create HMAC-SHA256 with key "WebAppData" and bot token
    4. Create HMAC-SHA256 with previous result and joined pairs
    5. Return as hex string
    """
    # Step 1: Create key=value pairs (excluding hash)
    pairs = []
    for key, value in data_dict.items():
        if key != 'hash':
            pairs.append(f"{key}={value}")

    # Step 2: Sort alphabetically
    pairs.sort()

    # Step 3: Create HMAC-SHA256 with "WebAppData" and bot token
    secret_key = hmac.new(
        "WebAppData".encode('utf-8'),
        bot_token.encode('utf-8'),
        hashlib.sha256
    ).digest()

    # Step 4: Create HMAC-SHA256 with secret key and joined pairs
    data_string = "\n".join(pairs)
    hash_value = hmac.new(
        secret_key,
        data_string.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()

    return hash_value

def generate_mock_init_data(bot_token: str,
                            user_data: Dict[str, Any] = None,
                            auth_date: int = None) -> str:
    """
    Generate complete mock Telegram init data string.

    Args:
        bot_token: Telegram bot token for hash generation
        user_data: Custom user data (uses MOCK_USER_DATA if None)
        auth_date: Unix timestamp for auth_date (uses current time if None)

    Returns:
        URL-encoded init data string
    """
    if user_data is None:
        user_data = MOCK_USER_DATA.copy()

    if auth_date is None:
        auth_date = int(time.time())

    # Create user JSON
    user_json = create_user_json(user_data)

    # Create init data parameters (without hash)
    init_data = {
        "user": user_json,
        "chat_type": "private",
        "auth_date": str(auth_date),
    }

    # Generate hash using the algorithm
    hash_value = generate_init_data_hash(init_data, bot_token)

    # Add hash to init data
    init_data["hash"] = hash_value

    # Convert to URL query string
    query_params = []
    for key, value in init_data.items():
        encoded_value = urllib.parse.quote(str(value), safe='')
        query_params.append(f"{key}={encoded_value}")

    return "&".join(query_params)

def parse_and_display_init_data(init_data_string: str) -> None:
    """
    Parse and display init data in a readable format.
    """
    parsed = urllib.parse.parse_qs(init_data_string)

    print("=== Parsed Init Data ===")
    for key, value_list in parsed.items():
        value = value_list[0]  # parse_qs returns lists

        if key == "user":
            try:
                # Try to parse and pretty-print user JSON
                user_data = json.loads(value)
                print(f"{key}:")
                for user_key, user_value in user_data.items():
                    print(f"  {user_key}: {user_value}")
            except json.JSONDecodeError:
                print(f"{key}: {value}")
        else:
            print(f"{key}: {value}")

def main():
    """
    Main function that generates mock init data once using bot token from dev.env
    """
    try:
        # Load bot token from dev.env file
        bot_token = load_bot_token()
        print("✅ Bot token loaded from dev.env")

        # Generate mock init data
        print("📝 Generating mock Telegram init data...")
        mock_init_data = generate_mock_init_data(bot_token)

        print("\n🎯 Generated Init Data:")
        print("=" * 80)
        print(mock_init_data)
        print("=" * 80)

        # Parse and display in readable format
        print("\n📊 Parsed Data:")
        print("-" * 50)
        parse_and_display_init_data(mock_init_data)

    except FileNotFoundError as e:
        print(f"❌ Error: {e}")
        print("\n💡 Create a dev.env file with:")
        print("BOT_TOKEN=your_telegram_bot_token_here")
    except ValueError as e:
        print(f"❌ Error: {e}")
    except Exception as e:
        print(f"❌ Unexpected error: {e}")

if __name__ == "__main__":
    main()

File: tests/initdata.go
package main

import (
	"bufio"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

// Mock user data similar to the Python version
var mockUserData = map[string]any{
	"id":                 687627953,
	"first_name":         "linuxfight",
	"last_name":          "",
	"username":           "linuxfight",
	"language_code":      "ru",
	"is_premium":         false,
	"allows_write_to_pm": false,
	"photo_url":          "https://example.com/cat.jpeg",
}

// loadBotToken reads BOT_TOKEN from dev.env (same format as the Python script)
func loadBotToken() (string, error) {
	file, err := os.Open("dev.env")
	if err != nil {
		return "", fmt.Errorf("dev.env not found: %w", err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if after, ok := strings.CutPrefix(line, "BOT_TOKEN="); ok {
			token := after
			token = strings.TrimSpace(token)
			if token == "" {
				return "", fmt.Errorf("BOT_TOKEN is empty in dev.env")
			}
			return token, nil
		}
	}
	if err := scanner.Err(); err != nil {
		return "", err
	}
	return "", fmt.Errorf("BOT_TOKEN not found in dev.env")
}

// createUserJSON removes empty string values (but keeps false booleans) and returns a compact JSON string.
func createUserJSON(userData map[string]any) (string, error) {
	filtered := make(map[string]any)
	for k, v := range userData {
		// Keep false booleans, remove empty strings, nil values are not expected here.
		if str, ok := v.(string); ok && str == "" {
			continue
		}
		filtered[k] = v
	}
	// Marshal without indentation and with compact separators.
	b, err := json.Marshal(filtered)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// generateHash replicates the algorithm from the Python script.
func generateHash(data map[string]string, botToken string) string {
	// Step 1: create key=value pairs excluding "hash"
	pairs := make([]string, 0, len(data))
	for k, v := range data {
		if k == "hash" {
			continue
		}
		pairs = append(pairs, fmt.Sprintf("%s=%s", k, v))
	}
	// Step 2: sort alphabetically
	sort.Strings(pairs)
	// Step 3: HMAC‑SHA256 with key "WebAppData" and bot token
	secretKey := hmac.New(sha256.New, []byte(botToken))
	secretKey.Write([]byte("WebAppData"))
	secret := secretKey.Sum(nil)
	// Step 4: HMAC‑SHA256 with secret key and joined pairs
	dataString := strings.Join(pairs, "\n")
	h := hmac.New(sha256.New, secret)
	h.Write([]byte(dataString))
	return fmt.Sprintf("%x", h.Sum(nil))
}

// generateMockInitData builds a mock init data string similar to the Python version.
func generateMockInitData(botToken string, userData map[string]any, authDate int64) (string, error) {
	if userData == nil {
		userData = mockUserData
	}
	if authDate == 0 {
		authDate = time.Now().Unix()
	}
	userJSON, err := createUserJSON(userData)
	if err != nil {
		return "", err
	}
	// Prepare init data without hash
	initData := map[string]string{
		"user":      userJSON,
		"chat_type": "private",
		"auth_date": strconv.FormatInt(authDate, 10),
	}
	// Generate hash
	hash := generateHash(initData, botToken)
	initData["hash"] = hash
	// Build URL‑encoded query string
	values := url.Values{}
	for k, v := range initData {
		values.Set(k, v)
	}
	return values.Encode(), nil
}

// parseAndDisplay prints the init data in a human‑readable form.
func parseAndDisplay(initDataString string) {
	parsed, _ := url.ParseQuery(initDataString)
	fmt.Println("=== Parsed Init Data ===")
	for key, vals := range parsed {
		value := vals[0]
		if key == "user" {
			var user map[string]any
			if err := json.Unmarshal([]byte(value), &user); err == nil {
				fmt.Printf("%s:\n", key)
				for uk, uv := range user {
					fmt.Printf("  %s: %v\n", uk, uv)
				}
				continue
			}
		}
		fmt.Printf("%s: %s\n", key, value)
	}
}

func main() {
	botToken, err := loadBotToken()
	if err != nil {
		fmt.Printf("❌ Error loading bot token: %v\n", err)
		fmt.Println("\n💡 Create a dev.env file with: BOT_TOKEN=your_telegram_bot_token_here")
		return
	}
	fmt.Println("✅ Bot token loaded from dev.env")
	mockInit, err := generateMockInitData(botToken, nil, 0)
	if err != nil {
		fmt.Printf("❌ Error generating mock init data: %v\n", err)
		return
	}
	fmt.Println("\n🎯 Generated Init Data:")
	fmt.Println(strings.Repeat("=", 80))
	fmt.Println(mockInit)
	fmt.Println(strings.Repeat("=", 80))
	fmt.Println("\n📊 Parsed Data:")
	fmt.Println(strings.Repeat("-", 50))
	parseAndDisplay(mockInit)
}
